### PR TITLE
Multi Modal Annotation v1.0 - "AutoTree"

### PR DESCRIPTION
**Pull request by hlydecker**
## Context

Actual code and work done by @mauch, PR ghost written by me.

Addressing #10 , and building on #5 + #7 , this PR adds functionality to automatically generate annotations in aerial imagery using the text prompt "tree".

## Features

- Exploration and testing of grounded segment anything geospatial
- `make_mask.py`: building on exploration done on grounded SAMGeo, this script provides basic WIP functionality for taking in input images and automatically generating masks to be fed through our conversion utility.

Merging this PR will solve much of #10

### CODE REVIEW

File: scripts/make_mask.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+        save_args (dict, optional): Save arguments for the prediction. Defaults to {}.
+        return_results (bool, optional): Whether to return the results. Defaults to False.
+
+    Returns:
+        tuple: Tuple containing masks, boxes, phrases, and logits.
+    """
+
+    if isinstance(image, str):
+        if image.startswith("http"):
+            image = download_file(image)
```
Comment (sahandv) [inline scripts/make_mask.py:103]: Quick question. Is `download_file` a standard library or something that has been missed to be included?

File: scripts/make_mask.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+        for box in self.boxes:
+            box = box.cpu().numpy()
+            boxlist.append((box[0], box[1]))
+        return boxlist
+
+
+def run_model(
+    input_images,
+    bt=0.23,
+    tt=0.24,
```
Comment (sahandv) [inline scripts/make_mask.py:196]: Are the names tt and bt standard? can't we just use `box_thresh` or `box_threshold` for instance?

File: scripts/make_mask.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        image_filename = os.path.basename(image)
+        output_basename = os.path.splitext(image_filename)[0]
+        annotate_trees(
+            image, output_root=os.path.join(output_dir, output_basename), **kwargs
+        )
+
+
+def annotate_trees(
+    input_image,
+    b_thresh=0.23,
```
Comment (sahandv) [inline scripts/make_mask.py:261]: `b_thresh`... can we make it `box_thresh` or `box_threshold`? Overall, make these standard across the code.

File: scripts/make_mask.py (hunk 1 above)
Comment (mauch) [reply scripts/make_mask.py:113]: `download_file` comes from `segment-geospatial` - I skipped the import because this function is just being monkey-patched into the `LangSAM` class to add the `box_reject` functionality and the import happens there.

I've added the import here in this module as well so that the linters don't complain.

File: scripts/make_mask.py (hunk 2 above)
Comment (mauch) [reply scripts/make_mask.py]: Yes - good point. I've changed it averywhere as you suggested.

File: scripts/make_mask.py (hunk 3 above)
Comment (mauch) [reply scripts/make_mask.py]: Same. I've changed it everywhere to `box_threshold`.

**Comment by mauch** [comment]
@sahandv  Can you please take another look? I've tidied things up quite a bit and added some extra functionality since you last reviewed.
Thanks.

File: scripts/make_mask.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+
+
+def is_empty(path):
+    """Check if specified path is a valid and empty dir."""
+    if os.path.exists(path) and not os.path.isfile(path):
+        # Checking if the directory is empty or not
+        if not os.listdir(path):
+            return True
+        else:
+            return False
```
Comment (xinwei89) [inline scripts/make_mask.py:29]: Maybe just `return not os.listdir(path)`?

File: scripts/make_mask.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    if output:
+        plt.savefig(output)
+    else:
+        plt.show()
+    plt.clf()
+    plt.close(fig)
+
+
+def predict_with_box_reject(
+    self,
```
Comment (xinwei89) [inline scripts/make_mask.py:67]: hmmm, maybe don't need the `self` keyword here?

File: scripts/make_mask.py [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+    box_threshold=0.23,
+    text_threshold=0.24,
+    output_dir="masks",
+    text_prompt="tree",
+    box_reject=0.99,
+):
+    """Generate a LangSAM segment anything geospatial model and predict on the
+    input images.
+
+    Parameters:
```
Comment (xinwei89) [inline scripts/make_mask.py:214]: Can we add some explanation maybe in the docstrings to explain the default values for the thresholds?

File: scripts/make_mask.py (hunk 4 above)
Comment (xinwei89) [reply scripts/make_mask.py:29]: Or maybe ` return os.path.exists(path) and os.path.isdir(path) and not os.listdir(path)` for the whole function  🤣

Comment (mauch) [reply scripts/make_mask.py]: Yes - good point. In fact I can't even remember what I was trying to do there. Looking at where I used it it appears that
its being used to check that the input file is a non empty directory so I think even just `return os.path.isdir(path) and not os.listdir(path)` will do the trick. I've changed it to that.

File: scripts/make_mask.py (hunk 6 above)
Comment (mauch) [reply scripts/make_mask.py]: OK sure. I've added a brief explanation.

**Review by xinwei89 — APPROVED** [review]
